Pinta.Gui.Widgets/Widgets/Layers/LayersListViewItemWidget.cs

```diff
@@ -106,57 +107,93 @@ public void HandleVisibilityToggled (bool visible)
 
 public sealed class LayersListViewItemWidget : Gtk.Box
 {
-	private static readonly Cairo.Pattern transparent_pattern = CairoExtensions.CreateTransparentBackgroundPattern (8);
+	private static readonly Pattern transparent_pattern = CairoExtensions.CreateTransparentBackgroundPattern (8);
 
 	private LayersListViewItem? item;
-	private Cairo.ImageSurface? thumbnail_surface;
+	private ImageSurface? thumbnail_surface;
 
 	private readonly Gtk.DrawingArea item_thumbnail;
 	private readonly Gtk.Label item_label;
 	private readonly Gtk.CheckButton visible_button;
 
 	public LayersListViewItemWidget ()
 	{
-		Spacing = 6;
+		Gtk.DrawingArea itemThumbnail = Gtk.DrawingArea.New ();
+		itemThumbnail.SetDrawFunc ((area, context, width, height) => DrawThumbnail (context, width, height));
+		itemThumbnail.WidthRequest = 60;
+		itemThumbnail.HeightRequest = 40;
+
+		Gtk.Label itemLabel = Gtk.Label.New (string.Empty);
+		itemLabel.Halign = Gtk.Align.Start;
+		itemLabel.Hexpand = true;
+		itemLabel.Ellipsize = Pango.EllipsizeMode.End;
+
+		Gtk.CheckButton visibleButton = Gtk.CheckButton.New ();
+		visibleButton.Halign = Gtk.Align.End;
+		visibleButton.Hexpand = false;
+		visibleButton.OnToggled += (_, _) => item?.HandleVisibilityToggled (visibleButton.Active);
+
+		Gtk.GestureClick menuGesture = Gtk.GestureClick.New ();
+		menuGesture.SetButton (Gdk.Constants.BUTTON_SECONDARY);
+		menuGesture.OnPressed += MenuGesture_OnPressed;
+
+		// --- Initialization (Gtk.Widget)
 
 		this.SetAllMargins (2);
+		this.AddController (menuGesture);
+
+		// --- Initialization (Gtk.Box)
+
+		Spacing = 6;
 
 		SetOrientation (Gtk.Orientation.Horizontal);
 
-		item_thumbnail = CreateItemThumbnail ();
-		item_label = CreateItemLabel ();
-		visible_button = CreateVisibleButton ();
+		Append (itemThumbnail);
+		Append (itemLabel);
+		Append (visibleButton);
 
-		Append (item_thumbnail);
-		Append (item_label);
-		Append (visible_button);
-	}
+		// --- References to keep
 
-	private Gtk.CheckButton CreateVisibleButton ()
-	{
-		Gtk.CheckButton result = Gtk.CheckButton.New ();
-		result.Halign = Gtk.Align.End;
-		result.Hexpand = false;
-		result.OnToggled += (_, _) => item?.HandleVisibilityToggled (visible_button.Active);
-		return result;
+		item_thumbnail = itemThumbnail;
+		item_label = itemLabel;
+		visible_button = visibleButton;
 	}
 
-	private static Gtk.Label CreateItemLabel ()
+	private void MenuGesture_OnPressed (
+		Gtk.GestureClick _,
+		Gtk.GestureClick.PressedSignalArgs args)
 	{
-		Gtk.Label result = Gtk.Label.New (string.Empty);
-		result.Halign = Gtk.Align.Start;
-		result.Hexpand = true;
-		result.Ellipsize = Pango.EllipsizeMode.End;
-		return result;
-	}
+		// TODO: Select layer before applying action
```

Comment:
The selection seemed to be happening okay in my testing (I assume from an event handler to the `SetCurrentUserLayer()` below?), but I'm not sure if the comment is describing a different issue?